### DISCUSSION TITLE
4.1.0

### DISCUSSION THREAD

**Discussion by SlicedSilver** [tradingview/lightweight-charts#1434] [General]
## Intro
Version 4.1 of Lightweight Charts introduces exciting new features, including the introduction of Plugins, which provide developers the ability to extend the library's functionality. Additionally, this release includes enhancements to customize the horizontal scale and various minor improvements and bug fixes.

### Plugins

Developers can now leverage the power of Plugins in Lightweight Charts. Two types of Plugins are supported -  [Custom Series](https://tradingview.github.io/lightweight-charts/docs/plugins/intro#custom-series) and [Drawing Primitives](https://tradingview.github.io/lightweight-charts/docs/plugins/intro#drawing-primitives), offering the ability to define new series types and create custom visualizations, drawing tools, and annotations.

With the flexibility provided by these plugins, developers can create highly customizable charting applications for their users.

To get started with plugins, please refer to our [Plugins Documentation](https://tradingview.github.io/lightweight-charts/docs/plugins/intro) for a better understanding of what is possible and how plugins work. You can also explore our collection of [plugin examples](https://github.com/tradingview/lightweight-charts/tree/master/plugin-examples) (with a [preview hosted here](https://tradingview.github.io/lightweight-charts/plugin-examples/)) for inspiration and guidance on implementing specific functionality.

To help you get started quickly, we have created an NPM package called [create-lwc-plugin](https://www.npmjs.com/package/create-lwc-plugin), which sets up a plugin project for you. This way, you can hit the ground running with your plugin development.

### Horizontal Scale Customization

The horizontal scale is no longer restricted to only time-based values. The API has been extended to allow customization of the horizontal scale behavior, and enable uses cases like options chart where price values are displayed in the horizontal scale. The most common use-case would be to customise the tick marks behaviour.

The [createChartEx](https://tradingview.github.io/lightweight-charts/docs/api#createchartex) function should be used instead of the usual `createChart` function, and an instance of a class implementing [IHorzScaleBehavior](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/IHorzScaleBehavior) should be provided.

A simple example can be found in this test case: [horizontal-price-scale.js](https://github.com/tradingview/lightweight-charts/blob/master/tests/e2e/graphics/test-cases/horizontal-price-scale.js)

### Enhancements

- Added point markers styling option for line-based series. (closes #365) [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/LineStyleOptions#pointmarkersvisible)
- Added double click subscriber for the main chart pane. (closes #1385) [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/IChartApi#subscribedblclick)
- Added ﻿`setCrosshairPosition` API, allowing programmatic setting of the crosshair position. (fixes #1198, #1163, #438) [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/IChartApi#setCrosshairPosition)
- Added an option to disable crosshair. Introduced the ﻿`Hidden` option in the ﻿`CrosshairMode` setting. (closes #749, thanks to @luk707) [Docs]([Docs](https://tradingview.github.io/lightweight-charts/docs/api/enums/CrosshairMode#hidden))
- Allow overriding tick mark label length via the ﻿`tickMarkMaxCharacterLength` option. (closes #1396) [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/HorzScaleOptions#tickmarkmaxcharacterlength)
- Support for overriding the percentage formatter within the localization options. (fixes #1328, #1291) [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/LocalizationOptions#percentageformatter)
- Added ﻿`paneSize` getter to `IChartApi`, returning the dimensions of the chart pane. (issue #1411) [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/IChartApi#panesize)
- Added options to set minimum dimensions for the price and time scales. (closes #1062, related to #1163, #50) [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/TimeScaleOptions#minimumheight), [Docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/PriceScaleOptions#minimumwidth)

### Bug Fixes

- Fixed chart layout when direction is set to ﻿RTL. (PR #1338)
- Fixed re-enabling of `autoSize` after disabling it. (PR #1377)
- Corrected percentage mode and zero first value. (fixes #1386)
- Prevent chart shifting when new data replaces existing whitespace. (fixes #1201)

Thanks to our Contributors for this Release:
- @luk707

You can always send us your feedback via GitHub.  
We look forward to hearing from you! And as always, happy trading!  

Team TradingView

See [issues assigned to this version's milestone](https://github.com/tradingview/lightweight-charts/milestone/24?closed=1) or [changes since the last published version](https://github.com/tradingview/lightweight-charts/compare/v4.0.1..v4.1.0).

<hr /><em>This discussion was created from the release <a href='https://github.com/tradingview/lightweight-charts/releases/tag/v4.1.0'>4.1.0</a>.</em>

**Comment by dokocat** [comment]
Very nice! @SlicedSilver [Custom Series](https://tradingview.github.io/lightweight-charts/docs/plugins/intro#custom-series) and [Drawing Primitives](https://tradingview.github.io/lightweight-charts/docs/plugins/intro#drawing-primitives) links are broken btw

**Reply by SlicedSilver** [reply]
You are too fast 😅
We deploy the website after a release.

**Reply by dokocat** [reply]
i live for these updates sir

**Reply by dokocat** [reply]
btw is it possible to do these https://tradingview.github.io/lightweight-charts/plugin-examples/plugins/expiring-price-alerts/example/ on regular tradingview charts?

**Reply by SlicedSilver** [reply]
It is not currently possible to do 'expiring price alerts' which look like that on TradingView charts.

**Reply by SlicedSilver** [reply]
FYI: updated the [demo](https://tradingview.github.io/lightweight-charts/plugin-examples/plugins/expiring-price-alerts/example/) of that plugin. Noticed that the auto scrolling of the chart wasn't working. We recently changed that behaviour because we believed it was bug. I forgot that example was using that 'bug' 😅.

**Reply by dokocat** [reply]
Wonderful! Thank you very much! Really appreciate your work on the plugin system, long time coming :)

**Reply by dokocat** [reply]
One more question! Is it now possible to achieve this with the plugin system? https://github.com/tradingview/lightweight-charts/discussions/1392

**Reply by SlicedSilver** [reply]
> One more question! Is it now possible to achieve this with the plugin system? #1392

Added a response 👍

**Comment by sarendipitee** [comment]
@SlicedSilver et al 

THIS IS HUGE. 💯 💯 💯 💯  THANK YOU SO MUCH. ❤️ 

I've been maintaining my own fork of LW-C in order to inject my own custom renderers and it's been a bit of a nightmare but now I can finally switch to upstream and maintain the renderers independently. This is everything I've been wanting for what is literally years now! Massive cheers and kudos!
